# Pandas


Pandas is a library built using NumPy specifically for data analysis. You'll be using Pandas heavily for data manipulation, visualisation, building machine learning models, etc. 

There are two main data structures in Pandas - Series and Dataframes. The default way to store data is dataframes, and thus manipulating dataframes quickly is probably the most important skill set for data analysis. 

*Source: https://pandas.pydata.org/pandas-docs/stable/overview.html*

In this section, you will study:
1. The pandas Series (similar to a numpy array)
    * Creating a pandas series
    * Indexing series
2. Dataframes 
    * Creating dataframes from dictionaries
    * Importing CSV data files as pandas dataframes
    * Reading and summarising dataframes
    * Sorting dataframes 

In [ ]:
pd.Series()



### 1. The Pandas Series 

A series is similar to a 1-D numpy array, and contains scalar values of the same type (numeric, character, datetime etc.). 
A dataframe is simply a table where each column is a pandas series.


#### Creating Pandas Series

Series are one-dimensional array-like structures, though unlike numpy arrays, they often contain non-numeric data (characters, dates, time, booleans etc.)

You can create pandas series from array-like objects using ```pd.Series()```.

In [1]:
# import pandas, pd is an alias
import pandas as pd
import numpy as np

# Creating a numeric pandas series
s = pd.Series([2, 4, 5, 6, 9])
print(s)
print(type(s))

0    2
1    4
2    5
3    6
4    9
dtype: int64
<class 'pandas.core.series.Series'>


Note that each element in the Series has an index, and the index starts at 0 as usual.

In [2]:
s.values

array([2, 4, 5, 6, 9], dtype=int64)

In [3]:
s.index

RangeIndex(start=0, stop=5, step=1)

In [4]:
# creating a series of characters
# notice that the 'dtype' here is 'object'
char_series = pd.Series(['a', 'b', 'af'])
char_series

0     a
1     b
2    af
dtype: object

In [5]:
# creating a series of type datetime
date_series = pd.date_range(start = '11-09-2017', end = '12-12-2017')
date_series

DatetimeIndex(['2017-11-09', '2017-11-10', '2017-11-11', '2017-11-12',
               '2017-11-13', '2017-11-14', '2017-11-15', '2017-11-16',
               '2017-11-17', '2017-11-18', '2017-11-19', '2017-11-20',
               '2017-11-21', '2017-11-22', '2017-11-23', '2017-11-24',
               '2017-11-25', '2017-11-26', '2017-11-27', '2017-11-28',
               '2017-11-29', '2017-11-30', '2017-12-01', '2017-12-02',
               '2017-12-03', '2017-12-04', '2017-12-05', '2017-12-06',
               '2017-12-07', '2017-12-08', '2017-12-09', '2017-12-10',
               '2017-12-11', '2017-12-12'],
              dtype='datetime64[ns]', freq='D')

In [6]:
type(date_series)

pandas.core.indexes.datetimes.DatetimeIndex

In [7]:
s

0    2
1    4
2    5
3    6
4    9
dtype: int64

#### Indexing Series

Indexing series is exactly same as 1-D numpy arrays - index starts at 0.

In [8]:
# Indexing pandas series: Same as indexing 1-d numpy arrays or lists
# accessing the fourth element
s[3]



6

In [9]:
# accessing elements starting index = 2 till the end
s[2:]

2    5
3    6
4    9
dtype: int64

In [10]:
# accessing the second and the fourth elements
# note that s[1, 3] will not work, you need to pass the indices [1, 3] as a list inside the original []
s[[1, 3]]

1    4
3    6
dtype: int64

In [11]:
s[1, 3]

KeyError: 'key of type tuple not found and not a MultiIndex'

#### Explicitly specifying indices

You might have noticed that while creating a series, Pandas automatically indexes it from 0 to (n-1), n being the number of rows. But if we want, we can also explicitly set the index ourselves, using the ‘index’ argument while creating the series using `pd.Series()`

In [12]:
# Indexing explicitly
pd.Series([0, 1, 2], index = ['a', 'b', 'c'])

a    0
b    1
c    2
dtype: int64

In [13]:
# You can also give the index as a sequence or use functions to specify the index
# But always make sure that the number of elements in the index list is equal to the number of elements specified in the series

pd.Series(np.array(range(0,10))**2, index = range(0,10))

0     0
1     1
2     4
3     9
4    16
5    25
6    36
7    49
8    64
9    81
dtype: int32

In [14]:
# 
s1=pd.Series([0, 1, 2,3], index = ['a', 'b', 'c','c'])

In [15]:
s1['c']

c    2
c    3
dtype: int64

Usually, you will work with Series only as a part of dataframes. Let's study the basics of dataframes.

In [16]:
# 
pd.Series([0, 1, 2,3], index = ['a', 'b', 'c','c'])



a    0
b    1
c    2
c    3
dtype: int64

### The Pandas Dataframe

Dataframe is the most widely used data-structure in data analysis. It is a table with rows and columns, with rows having an index and columns having meaningful names.

#### Creating dataframes from dictionaries

There are various ways of creating dataframes, such as creating them from dictionaries, JSON objects, reading from txt, CSV files, etc. 

In [17]:
type({'name': ['Vinay', 'Kushal', 'Aman', 'Saif'], 
                   'age': [22, 25, 24, 28], 
                    'occupation': ['engineer', 'doctor', 'data analyst', 'teacher']})

dict

In [18]:
# keys become column names
df = pd.DataFrame({'name': ['Vinay', 'Kushal', 'Aman', 'Saif'], 
                   'age': [22, 25, 24, 28], 
                    'occupation': ['engineer', 'doctor', 'data analyst', 'teacher']})
df

name  age    occupation
0   Vinay   22      engineer
1  Kushal   25        doctor
2    Aman   24  data analyst
3    Saif   28       teacher

#### Importing CSV data files as pandas dataframes 

For the upcoming exercises, we will use a dataset of a retail store having details about the orders placed, customers, product details, sales, profits etc. 



In [19]:
# reading a CSV file as a dataframe
market_df = pd.read_csv("market_fact.csv")

In [20]:
type(market_df)

pandas.core.frame.DataFrame

Usually, dataframes are imported as CSV files, but sometimes it is more convenient to convert dictionaries 
into dataframes. For e.g. when the raw data is in a JSON format (which is not uncommon), you can easily convert it into a dictionary, and then into a dataframe. 

You will learn how to convert JSON objects to dataframes later.

#### Reading and Summarising Dataframes

After you import a dataframe, you'd want to quickly understand its structure, shape, meanings of rows and columns etc. Further, you may want to look at summary statistics - such as mean, percentiles etc.

In [21]:
# Looking at the top and bottom entries of dataframes
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

In [22]:
market_df.tail()

Ord_id  Prod_id   Ship_id    Cust_id      Sales  Discount  \
8394  Ord_5353   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08   
8395  Ord_5411   Prod_6  SHP_7555  Cust_1798   127.1600      0.10   
8396  Ord_5388   Prod_6  SHP_7524  Cust_1798   243.0500      0.02   
8397  Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03   
8398  Ord_5459   Prod_6  SHP_7628  Cust_1798   603.6900      0.00   

      Order_Quantity  Profit  Shipping_Cost  Product_Base_Margin  
8394              28  374.63           7.69                 0.59  
8395              20  -74.03           6.92                 0.37  
8396              39  -70.85           5.35                 0.40  
8397              23  565.34          30.00                 0.62  
8398              47  131.39           4.86                 0.38

Here, each row represents an order placed at a retail store. Notice the index associated with each row - starts at 0 and ends at 8398, implying that there were 8399 orders placed.

In [23]:
# Looking at the datatypes of each column
market_df.info()

# Note that each column is basically a pandas Series of length 8399
# The ID columns are 'objects', i.e. they are being read as characters
# The rest are numeric (floats or int)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8399 entries, 0 to 8398
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Ord_id               8399 non-null   object 
 1   Prod_id              8399 non-null   object 
 2   Ship_id              8399 non-null   object 
 3   Cust_id              8399 non-null   object 
 4   Sales                8399 non-null   float64
 5   Discount             8399 non-null   float64
 6   Order_Quantity       8399 non-null   int64  
 7   Profit               8399 non-null   float64
 8   Shipping_Cost        8399 non-null   float64
 9   Product_Base_Margin  8336 non-null   float64
dtypes: float64(5), int64(1), object(4)
memory usage: 656.3+ KB


In [24]:
# Describe gives you a summary of all the numeric columns in the dataset
market_df.describe()

Sales     Discount  Order_Quantity        Profit  Shipping_Cost  \
count   8399.000000  8399.000000     8399.000000   8399.000000    8399.000000   
mean    1775.878179     0.049671       25.571735    181.184424      12.838557   
std     3585.050525     0.031823       14.481071   1196.653371      17.264052   
min        2.240000     0.000000        1.000000 -14140.700000       0.490000   
25%      143.195000     0.020000       13.000000    -83.315000       3.300000   
50%      449.420000     0.050000       26.000000     -1.500000       6.070000   
75%     1709.320000     0.080000       38.000000    162.750000      13.990000   
max    89061.050000     0.250000       50.000000  27220.690000     164.730000   

       Product_Base_Margin  
count          8336.000000  
mean              0.512513  
std               0.135589  
min               0.350000  
25%               0.380000  
50%               0.520000  
75%               0.590000  
max               0.850000

In [ ]:
#50% of orders are lossy in nature (-1.50---- 50% profit)

In [25]:
# Column names
market_df.columns

Index(['Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Sales', 'Discount',
       'Order_Quantity', 'Profit', 'Shipping_Cost', 'Product_Base_Margin'],
      dtype='object')

In [26]:
# The number of rows and columns
market_df.shape

(8399, 10)

#### Indices 

An important concept in pandas dataframes is that of *row indices*. By default, each row is assigned indices starting from 0, and are represented at the left side of the dataframe. 

In [27]:
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

Now, arbitrary numeric indices are difficult to read and work with. Thus, you may want to change the indices of the df to something more meanigful.

Let's change the index to Ord_id (unique id of each order), so that you can select rows using the order ids directly.

In [28]:
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

In [29]:
market_df.drop('Sales',axis=1)

Ord_id  Prod_id   Ship_id    Cust_id  Discount  Order_Quantity  \
0     Ord_5446  Prod_16  SHP_7609  Cust_1818      0.01              23   
1     Ord_5406  Prod_13  SHP_7549  Cust_1818      0.01              13   
2     Ord_5446   Prod_4  SHP_7610  Cust_1818      0.00              26   
3     Ord_5456   Prod_6  SHP_7625  Cust_1818      0.09              43   
4     Ord_5485  Prod_17  SHP_7664  Cust_1818      0.08              35   
...        ...      ...       ...        ...       ...             ...   
8394  Ord_5353   Prod_4  SHP_7479  Cust_1798      0.08              28   
8395  Ord_5411   Prod_6  SHP_7555  Cust_1798      0.10              20   
8396  Ord_5388   Prod_6  SHP_7524  Cust_1798      0.02              39   
8397  Ord_5348  Prod_15  SHP_7469  Cust_1798      0.03              23   
8398  Ord_5459   Prod_6  SHP_7628  Cust_1798      0.00              47   

       Profit  Shipping_Cost  Product_Base_Margin  
0      -30.51           3.60                 0.56  
1        4.56           0.93                 0.54  
2     1148.90           2.50                 0.59  
3      729.34          14.30                 0.37  
4     1219.87          26.30                 0.38  
...       ...            ...                  ...  
8394   374.63           7.69                 0.59  
8395   -74.03           6.92                 0.37  
8396   -70.85           5.35                 0.40  
8397   565.34          30.00                 0.62  
8398   131.39           4.86                 0.38  

[8399 rows x 9 columns]

In [30]:
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

In [31]:
market_df=market_df.drop('Sales',axis=1)

In [32]:
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id  Discount  Order_Quantity   Profit  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818      0.01              23   -30.51   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818      0.01              13     4.56   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818      0.00              26  1148.90   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818      0.09              43   729.34   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818      0.08              35  1219.87   

   Shipping_Cost  Product_Base_Margin  
0           3.60                 0.56  
1           0.93                 0.54  
2           2.50                 0.59  
3          14.30                 0.37  
4          26.30                 0.38

In [33]:
market_df.drop('Profit',axis=1)

Ord_id  Prod_id   Ship_id    Cust_id  Discount  Order_Quantity  \
0     Ord_5446  Prod_16  SHP_7609  Cust_1818      0.01              23   
1     Ord_5406  Prod_13  SHP_7549  Cust_1818      0.01              13   
2     Ord_5446   Prod_4  SHP_7610  Cust_1818      0.00              26   
3     Ord_5456   Prod_6  SHP_7625  Cust_1818      0.09              43   
4     Ord_5485  Prod_17  SHP_7664  Cust_1818      0.08              35   
...        ...      ...       ...        ...       ...             ...   
8394  Ord_5353   Prod_4  SHP_7479  Cust_1798      0.08              28   
8395  Ord_5411   Prod_6  SHP_7555  Cust_1798      0.10              20   
8396  Ord_5388   Prod_6  SHP_7524  Cust_1798      0.02              39   
8397  Ord_5348  Prod_15  SHP_7469  Cust_1798      0.03              23   
8398  Ord_5459   Prod_6  SHP_7628  Cust_1798      0.00              47   

      Shipping_Cost  Product_Base_Margin  
0              3.60                 0.56  
1              0.93                 0.54  
2              2.50                 0.59  
3             14.30                 0.37  
4             26.30                 0.38  
...             ...                  ...  
8394           7.69                 0.59  
8395           6.92                 0.37  
8396           5.35                 0.40  
8397          30.00                 0.62  
8398           4.86                 0.38  

[8399 rows x 8 columns]

In [34]:
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id  Discount  Order_Quantity   Profit  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818      0.01              23   -30.51   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818      0.01              13     4.56   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818      0.00              26  1148.90   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818      0.09              43   729.34   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818      0.08              35  1219.87   

   Shipping_Cost  Product_Base_Margin  
0           3.60                 0.56  
1           0.93                 0.54  
2           2.50                 0.59  
3          14.30                 0.37  
4          26.30                 0.38

In [35]:
market_df.drop('Profit',axis=1,inplace=True) #market_df=market_df.drop('Profit',axis=1)

In [36]:
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818      0.08              35   

   Shipping_Cost  Product_Base_Margin  
0           3.60                 0.56  
1           0.93                 0.54  
2           2.50                 0.59  
3          14.30                 0.37  
4          26.30                 0.38

In [37]:
# Setting index to Ord_id
market_df.set_index('Ord_id', inplace = True)
market_df.head()

Prod_id   Ship_id    Cust_id  Discount  Order_Quantity  \
Ord_id                                                             
Ord_5446  Prod_16  SHP_7609  Cust_1818      0.01              23   
Ord_5406  Prod_13  SHP_7549  Cust_1818      0.01              13   
Ord_5446   Prod_4  SHP_7610  Cust_1818      0.00              26   
Ord_5456   Prod_6  SHP_7625  Cust_1818      0.09              43   
Ord_5485  Prod_17  SHP_7664  Cust_1818      0.08              35   

          Shipping_Cost  Product_Base_Margin  
Ord_id                                        
Ord_5446           3.60                 0.56  
Ord_5406           0.93                 0.54  
Ord_5446           2.50                 0.59  
Ord_5456          14.30                 0.37  
Ord_5485          26.30                 0.38

Having meaningful row labels as indices helps you to select (subset) dataframes easily. You will study selecting dataframes in the next section. 

#### Sorting dataframes

You can sort dataframes in two ways - 1) by the indices and 2) by the values.  


In [38]:
# Sorting by index
# axis = 0 indicates that you want to sort rows (use axis=1 for columns)
market_df.sort_index(axis = 0, ascending = False)

Prod_id   Ship_id   Cust_id  Discount  Order_Quantity  \
Ord_id                                                            
Ord_999   Prod_15  SHP_1383  Cust_361      0.00              33   
Ord_998    Prod_8  SHP_1380  Cust_372      0.00              33   
Ord_998    Prod_5  SHP_1382  Cust_372      0.03              42   
Ord_998    Prod_8  SHP_1381  Cust_372      0.01               8   
Ord_997   Prod_14  SHP_1379  Cust_365      0.04               8   
...           ...       ...       ...       ...             ...   
Ord_1001   Prod_5  SHP_1385  Cust_374      0.07              49   
Ord_1000   Prod_6  SHP_1384  Cust_373      0.01              25   
Ord_100    Prod_8   SHP_138   Cust_58      0.10               3   
Ord_10     Prod_3    SHP_13   Cust_10      0.02              15   
Ord_1      Prod_1     SHP_1    Cust_1      0.04               6   

          Shipping_Cost  Product_Base_Margin  
Ord_id                                        
Ord_999           30.00                 0.62  
Ord_998            4.00                 0.60  
Ord_998           19.99                 0.55  
Ord_998            6.50                 0.79  
Ord_997           24.49                 0.37  
...                 ...                  ...  
Ord_1001           8.66                 0.76  
Ord_1000           6.47                 0.38  
Ord_100            1.99                 0.44  
Ord_10             2.99                 0.37  
Ord_1             35.00                 0.80  

[8399 rows x 7 columns]

In [ ]:
# Sorting by values

# Sorting in increasing order of Sales
#market_df.sort_values(by='Sales').head()

In [39]:
# Sorting in decreasing order of Shipping_Cost
market_df.sort_values(by='Shipping_Cost', ascending = False).head()

Prod_id   Ship_id   Cust_id  Discount  Order_Quantity  \
Ord_id                                                            
Ord_1751  Prod_15  SHP_2426  Cust_597      0.00              46   
Ord_839   Prod_11  SHP_1361  Cust_364      0.04              44   
Ord_1741  Prod_11  SHP_2411  Cust_595      0.02              26   
Ord_417   Prod_11   SHP_561  Cust_156      0.02              45   
Ord_1581  Prod_15  SHP_2184  Cust_519      0.07              17   

          Shipping_Cost  Product_Base_Margin  
Ord_id                                        
Ord_1751         164.73                 0.56  
Ord_839          154.12                 0.76  
Ord_1741         147.12                 0.80  
Ord_417          147.12                 0.80  
Ord_1581         143.71                 0.55

In [40]:
# Sorting by more than two columns

# Sorting in ascending order of Sales for each Product
market_df.sort_values(by=['Prod_id', 'Shipping_Cost'], ascending = False)

Prod_id   Ship_id    Cust_id  Discount  Order_Quantity  \
Ord_id                                                            
Ord_2973  Prod_9  SHP_6073  Cust_1480      0.09              36   
Ord_1152  Prod_9  SHP_1586   Cust_443      0.00              16   
Ord_2853  Prod_9  SHP_3927  Cust_1056      0.10              16   
Ord_1793  Prod_9  SHP_2484   Cust_604      0.03               7   
Ord_4356  Prod_9  SHP_6074  Cust_1481      0.01              41   
...          ...       ...        ...       ...             ...   
Ord_3354  Prod_1  SHP_4650  Cust_1236      0.03               9   
Ord_3549  Prod_1  SHP_4915  Cust_1273      0.00              14   
Ord_3562  Prod_1  SHP_4932  Cust_1265      0.01              14   
Ord_4758  Prod_1  SHP_6637  Cust_1638      0.10              35   
Ord_5364  Prod_1  SHP_7493  Cust_1805      0.02              37   

          Shipping_Cost  Product_Base_Margin  
Ord_id                                        
Ord_2973          19.99                 0.39  
Ord_1152          19.99                 0.40  
Ord_2853          19.99                 0.39  
Ord_1793          19.99                 0.39  
Ord_4356          19.99                 0.39  
...                 ...                  ...  
Ord_3354           4.51                 0.59  
Ord_3549           4.51                 0.59  
Ord_3562           4.51                 0.59  
Ord_4758           4.51                 0.59  
Ord_5364           4.51                 0.59  

[8399 rows x 7 columns]